File: android/src/test/java/com/google/android/libraries/privacy/ppn/krypton/TimerManagerTest.java
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the );
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an  BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.android.libraries.privacy.ppn.krypton;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.doAnswer;
import static org.robolectric.Shadows.shadowOf;

import android.os.ConditionVariable;
import android.os.Looper;
import androidx.test.core.app.ApplicationProvider;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.work.WorkManager;
import androidx.work.testing.WorkManagerTestInitHelper;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.TimeUnit;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

@RunWith(AndroidJUnit4.class)
public final class TimerManagerTest {
  private static final String TAG = "TimerManagerTest";
  @Rule public final MockitoRule mocks = MockitoJUnit.rule();
  @Mock private TimerListener timerExpiryListener;
  private TimerIdManager timerIdManager;

  @Before
  public void setUp() {
    WorkManagerTestInitHelper.initializeTestWorkManager(
        ApplicationProvider.getApplicationContext());
    timerIdManager =
        new TimerIdManager(
            timerExpiryListener,
            (WorkManager)
                WorkManagerTestInitHelper.getTestDriver(
                    ApplicationProvider.getApplicationContext()));
  }

  @Test
  public void timerManagerStartAndExpiry_expectOnExpiry() throws Exception {
    // Set up a timer.
    final ConditionVariable condition = new ConditionVariable(false);
    doAnswer(
            invocation -> {
              condition.open();
              return null;
            })
        .when(timerExpiryListener)
        .onTimerExpired(1);

    // Start the timer to expire in 10ms.
    assertThat(timerIdManager.startTimer(1, 10)).isTrue();

    // Let the Looper run everything that will be ready in the next 15ms.
    shadowOf(Looper.getMainLooper()).idleFor(15, TimeUnit.MILLISECONDS);

    // Assert that the timer got run.
    assertThat(condition.block(1000)).isTrue();
  }

  @Test
  public void timerManagerStartAndCancel_expectTrueCancellation() throws Exception {
    assertThat(timerIdManager.startTimer(1, 1000)).isTrue();
    TimerIdTask task = timerIdManager.getTask(1);
    assertThat(timerIdManager.cancelTimer(1)).isTrue();
    // Verify that the manager called cancel on the task.
    assertThat(task.isCancelled()).isTrue();
    // Cancelling second time will result in error.
    assertThat(timerIdManager.cancelTimer(1)).isFalse();
    shadowOf(Looper.getMainLooper()).idle();
  }

  @Test
  public void timerManagerCancelAllTimers_expectAllCancellations() throws Exception {
    HashMap<Integer, TimerIdTask> tasks = new HashMap<>();
    for (int i = 0; i < 100; i++) {
      assertThat(timerIdManager.startTimer(i, 1000)).isTrue();
      tasks.put(i, timerIdManager.getTask(i));
    }
    assertThat(timerIdManager.size()).isEqualTo(100);
    timerIdManager.cancelAllTimers();
    assertThat(timerIdManager.size()).isEqualTo(0);
    for (Map.Entry<Integer, TimerIdTask> entry : tasks.entrySet()) {
      assertThat(entry.getValue().isCancelled()).isTrue();
    }
  }

  @Test
  public void timerManagerGetInstance_returnsCorrectManager() throws Exception {
    assertThat(TimerIdManager.getInstance(timerIdManager.getId())).isEqualTo(timerIdManager);
  }

  @Test
  public void timerManagerStop_removesManagerFromActiveManagers() throws Exception {
    UUID managerId = timerIdManager.getId();
    timerIdManager.stop();
    assertThat(TimerIdManager.getInstance(managerId)).isNull();
  }

  @Test
  public void timerManagerStop_cancelsRunningTimers() throws Exception {
    HashMap<Integer, TimerIdTask> tasks = new HashMap<>();
    for (int i = 0; i < 10; i++) {
      assertThat(timerIdManager.startTimer(i, 1000)).isTrue();
      tasks.put(i, timerIdManager.getTask(i));
    }
    assertThat(timerIdManager.size()).isEqualTo(10);
    timerIdManager.stop();
    assertThat(timerIdManager.size()).isEqualTo(0);
    for (Map.Entry<Integer, TimerIdTask> entry : tasks.entrySet()) {
      assertThat(entry.getValue().isCancelled()).isTrue();
    }
  }
}

File: krypton/timer_manager_test.cc
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "privacy/net/krypton/timer_manager.h"

#include <functional>

#include "privacy/net/krypton/pal/mock_timer_interface.h"
#include "testing/base/public/gmock.h"
#include "testing/base/public/gunit.h"
#include "third_party/absl/functional/bind_front.h"
#include "third_party/absl/status/status.h"
#include "third_party/absl/status/statusor.h"
#include "third_party/absl/synchronization/notification.h"
#include "third_party/absl/time/time.h"

namespace privacy {
namespace krypton {
namespace {

using ::testing::Return;
using ::testing::status::IsOkAndHolds;

class TimerManagerTest : public ::testing::Test {
 public:
  int StartATimer(int expected_timer_id, std::function<void()> callback) {
    // Expect a timer id of 0.
    EXPECT_CALL(timer_interface_,
                StartTimer(expected_timer_id, absl::Milliseconds(5)))
        .WillOnce(Return(absl::OkStatus()));
    EXPECT_THAT(timer_manager_.StartTimer(absl::Milliseconds(5), callback),
                IsOkAndHolds(expected_timer_id));
    return expected_timer_id;
  }

  MockTimerInterface timer_interface_;
  TimerManager timer_manager_{&timer_interface_};
};

// A simple object simulating a caller for timer manager.
struct FakeTimer {
  absl::Notification notification;
  void Callback() { notification.Notify(); }
  int timer_id = -1;
};

TEST_F(TimerManagerTest, TestStartAndExpiry) {
  FakeTimer foo;

  foo.timer_id = StartATimer(0, absl::bind_front(&FakeTimer::Callback, &foo));
  EXPECT_EQ(1, timer_manager_.NumActiveTimers());

  timer_interface_.TimerExpiry(foo.timer_id);
  foo.notification.WaitForNotification();
  EXPECT_EQ(0, timer_manager_.NumActiveTimers());
}

TEST_F(TimerManagerTest, TestStartAndCancel) {
  FakeTimer foo;

  foo.timer_id = StartATimer(0, absl::bind_front(&FakeTimer::Callback, &foo));
  EXPECT_EQ(1, timer_manager_.NumActiveTimers());

  EXPECT_CALL(timer_interface_, CancelTimer(foo.timer_id));
  timer_manager_.CancelTimer(foo.timer_id);
  EXPECT_EQ(0, timer_manager_.NumActiveTimers());
}

TEST_F(TimerManagerTest, TestMultipleTimersWithExpiry) {
  FakeTimer foo;
  FakeTimer bar;
  foo.timer_id = StartATimer(0, absl::bind_front(&FakeTimer::Callback, &foo));
  bar.timer_id = StartATimer(1, absl::bind_front(&FakeTimer::Callback, &bar));
  EXPECT_EQ(2, timer_manager_.NumActiveTimers());
  // Foo timer expired.
  timer_interface_.TimerExpiry(foo.timer_id);
  foo.notification.WaitForNotification();

  // Bar timer expired.
  timer_interface_.TimerExpiry(bar.timer_id);
  bar.notification.WaitForNotification();
  EXPECT_EQ(0, timer_manager_.NumActiveTimers());
}

}  // namespace
}  // namespace krypton
}  // namespace privacy
